subprojects/base-services/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java

```diff
@@ -200,12 +200,27 @@ int sizeOfCommonPrefix(String path, int offset) {
             return lastSeparator;
         }
 
+
+        /**
+         * This uses an optimized version of {@link String#regionMatches(int, String, int, int)}
+         * which does not check for negative indices or integer overflow.
+         */
         boolean isChildOfOrThis(String filePath, int offset) {
-            if (!filePath.regionMatches(offset, prefix, 0, prefix.length())) {
+            char[] pathChars = filePath.toCharArray();
+            char[] prefixChars = prefix.toCharArray();
```

Comment:
Instead of allocating a new char[] array, it should be more efficient to loop through the strings and grabbing characters using `String#charAt(int)`.

This looks great!

Reply:
Of course, how did I miss that? :)